src/datomish/clauses.cljc

```diff
@@ -0,0 +1,256 @@
+;; This Source Code Form is subject to the terms of the Mozilla Public
+;; License, v. 2.0. If a copy of the MPL was not distributed with this
+;; file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+(ns datomish.clauses
+  (:require
+     [datomish.source
+      :refer [attribute-in-source
+              constant-in-source
+              source->from
+              source->constraints]]
+     [datomish.util :as util #?(:cljs :refer-macros :clj :refer) [raise-str cond-let]]
+     [datascript.parser :as dp
+      #?@(:cljs
+            [:refer
+             [PlainSymbol Predicate Not Pattern DefaultSrc Variable Constant Placeholder]])]
+     [honeysql.core :as sql]
+     [clojure.string :as str]
+     )
+  #?(:clj
+       (:import
+          [datascript.parser
+           PlainSymbol Predicate Not Pattern DefaultSrc Variable Constant Placeholder])))
+
+;; A CC is a collection of clauses that are combined with JOIN.
+;; The topmost form in a query is a ConjoiningClauses.
+;;
+;; Ordinary pattern clauses turn into FROM parts and WHERE parts using :=.
+;; Predicate clauses turn into the same, but with other functions.
+;; Function clauses with bindings turn into:
+;;   * Subqueries. Perhaps less efficient? Certainly clearer.
```

Comment:
Can you clarify what's real and what's speculative here?  What's done and what's planned?